File: apps/scraper/app/hosts.py
import json
import logging
import re
from urllib.parse import urlparse

from bs4 import BeautifulSoup

from .parsing import ExtractionResult, _extract_price_from_text, _is_element_disabled, _parse_price_match

logger = logging.getLogger(__name__)


def _price_from_selectors(soup: BeautifulSoup, selectors: list[str]) -> tuple[float | None, str | None]:
    """Iterate CSS selectors, return the first price found."""
    for selector in selectors:
        for el in soup.select(selector):
            content = el.get("content")
            if content:
                try:
                    val = float(content)
                    if val > 0:
                        return val, f"price={val} ({selector} content)"
                except ValueError:
                    pass
            text = el.get_text(strip=True)
            extracted = _extract_price_from_text(text)
            if extracted is not None:
                return extracted, f"price={extracted} ({selector})"
    return None, None


def _extract_host_specific(url: str, soup: BeautifulSoup) -> ExtractionResult | None:
    hostname = (urlparse(url).hostname or "").lower()
    if "target." in hostname:
        return _extract_target_data(soup)
    if "amazon." in hostname:
        return _extract_amazon_data(soup)
    if "walmart." in hostname:
        return _extract_walmart_data(soup)
    if "bestbuy." in hostname:
        return _extract_bestbuy_data(soup)
    if "costco." in hostname:
        return _extract_costco_data(soup)
    return None


def _extract_target_data(soup: BeautifulSoup) -> ExtractionResult | None:
    """Extract price and stock from Target hydration/config scripts."""
    price = None
    stock_status = None
    raw_parts: list[str] = []

    price_patterns = [
        r'"formatted_current_price"\s*:\s*"([^"]+)"',
        r'"current_retail"\s*:\s*"?(?P<price>\d+(?:\.\d{1,2})?)"?',
        r'"retail"\s*:\s*"?(?P<price>\d+(?:\.\d{1,2})?)"?',
    ]
    stock_patterns = [
        (r'"availability_status"\s*:\s*"OUT_OF_STOCK"', "out_of_stock", "availability_status=OUT_OF_STOCK"),
        (r'"availability_status"\s*:\s*"IN_STOCK"', "in_stock", "availability_status=IN_STOCK"),
        (r'"is_out_of_stock"\s*:\s*true', "out_of_stock", "is_out_of_stock=true"),
        (r'"is_out_of_stock"\s*:\s*false', "in_stock", "is_out_of_stock=false"),
        (r'"available_to_promise_network"\s*:\s*"NOT_AVAILABLE"', "out_of_stock", "available_to_promise_network=NOT_AVAILABLE"),
        (r'"available_to_promise_network"\s*:\s*"AVAILABLE"', "in_stock", "available_to_promise_network=AVAILABLE"),
    ]

    for script in soup.find_all("script"):
        content = script.string or script.get_text() or ""
        if not content:
            continue
        if not any(
            marker in content
            for marker in (
                "current_retail",
                "formatted_current_price",
                "availability_status",
                "available_to_promise_network",
                "is_out_of_stock",
            )
        ):
            continue

        if price is None:
            for pattern in price_patterns:
                match = re.search(pattern, content)
                if not match:
                    continue

                if "formatted_current_price" in pattern:
                    extracted_price = _extract_price_from_text(match.group(1))
                else:
                    extracted_price = float(match.group("price"))

                if extracted_price is not None:
                    price = extracted_price
                    raw_parts.append(f"price={price}")
                    break

        if stock_status is None:
            for pattern, mapped_value, raw in stock_patterns:
                if re.search(pattern, content):
                    stock_status = mapped_value
                    raw_parts.append(raw)
                    break

        if price is not None and stock_status is not None:
            break

    if price is None and stock_status is None:
        return None

    return ExtractionResult(price=price, stock_status=stock_status, raw=", ".join(raw_parts))


def _extract_amazon_data(soup: BeautifulSoup) -> ExtractionResult | None:
    """Extract price and stock from Amazon product pages."""
    raw_parts: list[str] = []

    price_selectors = [
        "#corePriceDisplay_desktop_feature_div .aok-offscreen",
        ".a-price .a-offscreen",
        "#priceblock_ourprice",
        "#priceblock_dealprice",
        'span.a-price[data-a-color="price"] .a-offscreen',
        "#price_inside_buybox",
        "#newBuyBoxPrice",
    ]
    price, price_raw = _price_from_selectors(soup, price_selectors)
    if price_raw:
        raw_parts.append(price_raw)

    stock_status = None
    availability_el = soup.select_one("#availability")
    if availability_el:
        text = availability_el.get_text(strip=True).lower()
        if "in stock" in text:
            stock_status = "in_stock"
            raw_parts.append("availability=in_stock")
        elif "unavailable" in text or "out of stock" in text:
            stock_status = "out_of_stock"
            raw_parts.append("availability=out_of_stock")

    if stock_status is None:
        add_to_cart = soup.select_one("#add-to-cart-button")
        if add_to_cart and not _is_element_disabled(add_to_cart):
            stock_status = "in_stock"
            raw_parts.append("add-to-cart=present")

    if price is None and stock_status is None:
        return None

    return ExtractionResult(price=price, stock_status=stock_status, raw=", ".join(raw_parts))


def _extract_walmart_data(soup: BeautifulSoup) -> ExtractionResult | None:
    """Extract price and stock from Walmart product pages."""
    raw_parts: list[str] = []

    price_selectors = [
        '[data-testid="price-wrap"] [itemprop="price"]',
        'span[itemprop="price"]',
        '[data-testid="price-wrap"]',
    ]
    price, price_raw = _price_from_selectors(soup, price_selectors)
    if price_raw:
        raw_parts.append(price_raw)

    # Try __NEXT_DATA__ JSON
    if price is None:
        for script in soup.find_all("script", id="__NEXT_DATA__"):
            try:
                data = json.loads(script.string or "")
                props = data.get("props", {}).get("pageProps", {})
                item = props.get("initialData", {}).get("data", {}).get("product", {})
                price_info = item.get("priceInfo", {}).get("currentPrice", {})
                p = price_info.get("price")
                if p is not None:
                    price = float(p)
                    raw_parts.append(f"price={price} (__NEXT_DATA__)")
            except (json.JSONDecodeError, TypeError, ValueError, AttributeError):
                pass

    stock_status = None
    fulfillment = soup.select_one('[data-testid="fulfillment-badge"]')
    if fulfillment:
        text = fulfillment.get_text(strip=True).lower()
        if "out of stock" in text or "unavailable" in text:
            stock_status = "out_of_stock"
            raw_parts.append("fulfillment=out_of_stock")
        else:
            stock_status = "in_stock"
            raw_parts.append("fulfillment=in_stock")

    if price is None and stock_status is None:
        return None

    return ExtractionResult(price=price, stock_status=stock_status, raw=", ".join(raw_parts))


def _extract_bestbuy_data(soup: BeautifulSoup) -> ExtractionResult | None:
    """Extract price and stock from Best Buy product pages."""
    raw_parts: list[str] = []

    price_selectors = [
        ".priceView-hero-price .priceView-customer-price span:first-child",
        "[data-testid='customer-price'] span",
        ".priceView-hero-price span",
    ]
    price, price_raw = _price_from_selectors(soup, price_selectors)
    if price_raw:
        raw_parts.append(price_raw)

    stock_status = None
    sold_out = soup.select_one("[data-button-state='SOLD_OUT']")
    if sold_out:
        stock_status = "out_of_stock"
        raw_parts.append("button=SOLD_OUT")
    else:
        add_to_cart = soup.select_one(".fulfillment-add-to-cart-button")
        if add_to_cart and not _is_element_disabled(add_to_cart):
            stock_status = "in_stock"
            raw_parts.append("add-to-cart=present")

    if price is None and stock_status is None:
        return None

    return ExtractionResult(price=price, stock_status=stock_status, raw=", ".join(raw_parts))


def _extract_costco_data(soup: BeautifulSoup) -> ExtractionResult | None:
    """Extract price and stock from Costco product pages."""
    raw_parts: list[str] = []

    price_selectors = [
        "#pull-right-price span.value",
        ".your-price .value",
        "[automation-id='productPrice']",
    ]
    price, price_raw = _price_from_selectors(soup, price_selectors)
    if price_raw:
        raw_parts.append(price_raw)

    stock_status = None
    page_text = soup.get_text(strip=True).lower()
    if "out of stock" in page_text:
        stock_status = "out_of_stock"
        raw_parts.append("text=out_of_stock")
    else:
        add_to_cart = soup.select_one("#add-to-cart-btn, #addToCartButton, .add-to-cart-btn")
        if add_to_cart and not _is_element_disabled(add_to_cart):
            stock_status = "in_stock"
            raw_parts.append("add-to-cart=present")

    if price is None and stock_status is None:
        return None

    return ExtractionResult(price=price, stock_status=stock_status, raw=", ".join(raw_parts))

File: apps/scraper-rs/src/hosts.rs
use regex::Regex;
use scraper::Html;
use serde_json::Value;
use url::Url;

use crate::{
    dom::{select_all, select_first},
    parsing::{
        ExtractionResult, extract_price_from_text, is_element_disabled, is_purchase_cta_text,
        parse_price_match, text_of_element,
    },
};

pub fn extract_host_specific(url: &str, document: &Html) -> Option<ExtractionResult> {
    let hostname = Url::parse(url)
        .ok()
        .and_then(|url| url.host_str().map(str::to_lowercase))
        .unwrap_or_default();

    if hostname.contains("target.") {
        return extract_target_data(document);
    }
    if hostname.contains("amazon.") {
        return extract_amazon_data(document);
    }
    if hostname.contains("walmart.") {
        return extract_walmart_data(document);
    }
    if hostname.contains("bestbuy.") {
        return extract_bestbuy_data(document);
    }
    if hostname.contains("costco.") {
        return extract_costco_data(document);
    }

    None
}

fn price_from_selectors(document: &Html, selectors: &[&str]) -> (Option<f64>, Option<String>) {
    for selector in selectors {
        for element in select_all(document, selector) {
            if let Some(content) = element.value().attr("content") {
                if let Some(price) = parse_price_match(content) {
                    return (
                        Some(price),
                        Some(format!("price={price} ({selector} content)")),
                    );
                }
            }

            let text = text_of_element(&element);
            if let Some(price) = extract_price_from_text(&text) {
                return (Some(price), Some(format!("price={price} ({selector})")));
            }
        }
    }

    (None, None)
}

fn extract_target_data(document: &Html) -> Option<ExtractionResult> {
    let price_patterns = [
        Regex::new(r#""formatted_current_price"\s*:\s*"([^"]+)""#).expect("valid target regex"),
        Regex::new(r#""current_retail"\s*:\s*"?(?P<price>\d+(?:\.\d{1,2})?)"?"#)
            .expect("valid target regex"),
        Regex::new(r#""retail"\s*:\s*"?(?P<price>\d+(?:\.\d{1,2})?)"?"#)
            .expect("valid target regex"),
    ];
    let stock_patterns = [
        (
            Regex::new(r#""availability_status"\s*:\s*"OUT_OF_STOCK""#)
                .expect("valid target regex"),
            "out_of_stock",
            "availability_status=OUT_OF_STOCK",
        ),
        (
            Regex::new(r#""availability_status"\s*:\s*"IN_STOCK""#).expect("valid target regex"),
            "in_stock",
            "availability_status=IN_STOCK",
        ),
        (
            Regex::new(r#""is_out_of_stock"\s*:\s*true"#).expect("valid target regex"),
            "out_of_stock",
            "is_out_of_stock=true",
        ),
        (
            Regex::new(r#""is_out_of_stock"\s*:\s*false"#).expect("valid target regex"),
            "in_stock",
            "is_out_of_stock=false",
        ),
        (
            Regex::new(r#""available_to_promise_network"\s*:\s*"NOT_AVAILABLE""#)
                .expect("valid target regex"),
            "out_of_stock",
            "available_to_promise_network=NOT_AVAILABLE",
        ),
        (
            Regex::new(r#""available_to_promise_network"\s*:\s*"AVAILABLE""#)
                .expect("valid target regex"),
            "in_stock",
            "available_to_promise_network=AVAILABLE",
        ),
    ];

    let mut price = None;
    let mut stock_status = None;
    let mut raw_parts = Vec::new();

    for script in select_all(document, "script") {
        let content = text_of_element(&script);
        if content.is_empty()
            || ![
                "current_retail",
                "formatted_current_price",
                "availability_status",
                "available_to_promise_network",
                "is_out_of_stock",
            ]
            .iter()
            .any(|marker| content.contains(marker))
        {
            continue;
        }

        if price.is_none() {
            for pattern in &price_patterns {
                if let Some(captures) = pattern.captures(&content) {
                    let extracted = captures
                        .name("price")
                        .map(|value| value.as_str().to_string())
                        .or_else(|| captures.get(1).map(|value| value.as_str().to_string()))
                        .and_then(|value| {
                            extract_price_from_text(&value).or_else(|| parse_price_match(&value))
                        });
                    if let Some(extracted) = extracted {
                        price = Some(extracted);
                        raw_parts.push(format!("price={extracted}"));
                        break;
                    }
                }
            }
        }

        if stock_status.is_none() {
            for (pattern, mapped_value, raw_value) in &stock_patterns {
                if pattern.is_match(&content) {
                    stock_status = Some((*mapped_value).to_string());
                    raw_parts.push((*raw_value).to_string());
                    break;
                }
            }
        }

        if price.is_some() && stock_status.is_some() {
            break;
        }
    }

    if price.is_none() && stock_status.is_none() {
        return None;
    }

    Some(ExtractionResult {
        price,
        stock_status,
        raw: Some(raw_parts.join(", ")),
    })
}

fn extract_amazon_data(document: &Html) -> Option<ExtractionResult> {
    let price_selectors = [
        "#corePriceDisplay_desktop_feature_div .aok-offscreen",
        ".a-price .a-offscreen",
        "#priceblock_ourprice",
        "#priceblock_dealprice",
        r#"span.a-price[data-a-color="price"] .a-offscreen"#,
        "#price_inside_buybox",
        "#newBuyBoxPrice",
    ];
    let (price, price_raw) = price_from_selectors(document, &price_selectors);
    let mut raw_parts = price_raw.into_iter().collect::<Vec<_>>();

    let mut stock_status = None;
    if let Some(availability) = select_first(document, "#availability") {
        let text = text_of_element(&availability).to_lowercase();
        if text.contains("in stock") {
            stock_status = Some("in_stock".to_string());
            raw_parts.push("availability=in_stock".to_string());
        } else if text.contains("unavailable") || text.contains("out of stock") {
            stock_status = Some("out_of_stock".to_string());
            raw_parts.push("availability=out_of_stock".to_string());
        }
    }

    if stock_status.is_none() {
        if let Some(add_to_cart) = select_first(document, "#add-to-cart-button") {
            let text = add_to_cart
                .value()
                .attr("value")
                .map(str::to_string)
                .unwrap_or_else(|| text_of_element(&add_to_cart));
            if is_purchase_cta_text(&text) {
                if is_element_disabled(&add_to_cart) {
                    stock_status = Some("out_of_stock".to_string());
                    raw_parts.push("add-to-cart=disabled".to_string());
                } else {
                    stock_status = Some("in_stock".to_string());
                    raw_parts.push("add-to-cart=present".to_string());
                }
            }
        }
    }

    if price.is_none() && stock_status.is_none() {
        return None;
    }

    Some(ExtractionResult {
        price,
        stock_status,
        raw: Some(raw_parts.join(", ")),
    })
}

fn extract_walmart_data(document: &Html) -> Option<ExtractionResult> {
    let price_selectors = [
        r#"[data-testid="price-wrap"] [itemprop="price"]"#,
        r#"span[itemprop="price"]"#,
        r#"[data-testid="price-wrap"]"#,
    ];
    let (mut price, price_raw) = price_from_selectors(document, &price_selectors);
    let mut raw_parts = price_raw.into_iter().collect::<Vec<_>>();

    if price.is_none() {
        for script in select_all(document, r#"script[id="__NEXT_DATA__"]"#) {
            let raw = text_of_element(&script);
            let Ok(data) = serde_json::from_str::<Value>(&raw) else {
                continue;
            };

            if let Some(value) = data
                .pointer("/props/pageProps/initialData/data/product/priceInfo/currentPrice/price")
                .and_then(Value::as_f64)
            {
                price = Some(value);
                raw_parts.push(format!("price={value} (__NEXT_DATA__)"));
                break;
            }
        }
    }

    let mut stock_status = None;
    if let Some(fulfillment) = select_first(document, r#"[data-testid="fulfillment-badge"]"#) {
        let text = text_of_element(&fulfillment).to_lowercase();
        if text.contains("out of stock") || text.contains("unavailable") {
            stock_status = Some("out_of_stock".to_string());
            raw_parts.push("fulfillment=out_of_stock".to_string());
        } else {
            stock_status = Some("in_stock".to_string());
            raw_parts.push("fulfillment=in_stock".to_string());
        }
    }

    if price.is_none() && stock_status.is_none() {
        return None;
    }

    Some(ExtractionResult {
        price,
        stock_status,
        raw: Some(raw_parts.join(", ")),
    })
}

fn extract_bestbuy_data(document: &Html) -> Option<ExtractionResult> {
    let price_selectors = [
        ".priceView-hero-price .priceView-customer-price span:first-child",
        "[data-testid='customer-price'] span",
        ".priceView-hero-price span",
    ];
    let (price, price_raw) = price_from_selectors(document, &price_selectors);
    let mut raw_parts = price_raw.into_iter().collect::<Vec<_>>();

    let mut stock_status = None;
    if select_first(document, r#"[data-button-state="SOLD_OUT"]"#).is_some() {
        stock_status = Some("out_of_stock".to_string());
        raw_parts.push("button=SOLD_OUT".to_string());
    } else if let Some(add_to_cart) = select_first(document, ".fulfillment-add-to-cart-button") {
        let text = text_of_element(&add_to_cart);
        if is_purchase_cta_text(&text) {
            if is_element_disabled(&add_to_cart) {
                stock_status = Some("out_of_stock".to_string());
                raw_parts.push("add-to-cart=disabled".to_string());
            } else {
                stock_status = Some("in_stock".to_string());
                raw_parts.push("add-to-cart=present".to_string());
            }
        }
    }

    if price.is_none() && stock_status.is_none() {
        return None;
    }

    Some(ExtractionResult {
        price,
        stock_status,
        raw: Some(raw_parts.join(", ")),
    })
}

fn extract_costco_data(document: &Html) -> Option<ExtractionResult> {
    let price_selectors = [
        "#pull-right-price span.value",
        ".your-price .value",
        r#"[automation-id="productPrice"]"#,
    ];
    let (price, price_raw) = price_from_selectors(document, &price_selectors);
    let mut raw_parts = price_raw.into_iter().collect::<Vec<_>>();

    let mut stock_status = None;
    let page_text = document.root_element().text().collect::<Vec<_>>().join(" ");
    if page_text.to_lowercase().contains("out of stock") {
        stock_status = Some("out_of_stock".to_string());
        raw_parts.push("text=out_of_stock".to_string());
    } else if let Some(add_to_cart) = select_first(
        document,
        "#add-to-cart-btn, #addToCartButton, .add-to-cart-btn",
    ) {
        let text = text_of_element(&add_to_cart);
        if is_purchase_cta_text(&text) {
            if is_element_disabled(&add_to_cart) {
                stock_status = Some("out_of_stock".to_string());
                raw_parts.push("add-to-cart=disabled".to_string());
            } else {
                stock_status = Some("in_stock".to_string());
                raw_parts.push("add-to-cart=present".to_string());
            }
        }
    }

    if price.is_none() && stock_status.is_none() {
        return None;
    }

    Some(ExtractionResult {
        price,
        stock_status,
        raw: Some(raw_parts.join(", ")),
    })
}
